arch/sim/src/sim/sim_alsa.c

```diff
@@ -31,17 +31,31 @@
 #include <debug.h>
 
 #include <alsa/asoundlib.h>
+#include <mad.h>
 
 /****************************************************************************
  * Pre-processor Definitions
  ****************************************************************************/
 
 #define AUDMIN(a,b)     ((a) > (b) ? (b) : (a))
+#define AUDCODEC_DEC    0x01
+#define AUDCODEC_ENC    0x10
 
 /****************************************************************************
  * Private Types
  ****************************************************************************/
 
+struct sim_codec_ops_s
+{
+  uint8_t format;
+  uint8_t flags;
+  void    *(*init)(void);
+  int      (*get_samples)(void *handle);
+  int      (*process)(void *handle, unsigned char *in, unsigned int insize,
+                      unsigned char *out, unsigned int *outsize);
```

Comment:
let's process return out buffer instead caller allocate out buffer before call process:
 unsigned char **out, unsigned int *outsize);
the benefit is that:

1. get_samples callback could be removed
2. pcm codec can save memcpy by  assign in to out directly
3. mp3 decode can reuse the same out buffer instead allocate every time.

Reply:
changed to `unsigned char **out`, and remove memcpy in pcm codec.  
but i think get_sample is necessary: 
1, when playback,   pcm from decoder wether write to alsa once time.
2, when capture,  available pcm of alsa  wether meet requirement of encoder. 
and each codec is different.

Reply:
encode/decode can save the partial internally?